File: clojure/src/ch1/ex119.clj
;; (load-file "filename.clj")

(defn sq [n]
  (* n n))

(defn fib-iter [a b p q count]
  (cond (= count 0) b
        (even? count)
          (fib-iter a
                    b
                    (+ (sq p) (sq q))
                    (+ (sq q) (* 2 p q))
                    (/ count 2))
        :else (fib-iter (+ (* b q) (* a q) (* a p))
                     (+ (* b p) (* a q))
                     p
                     q
                     (- count 1))))

(defn fib [n]
  (fib-iter 1 0 0 1 n))


(require
 'clojure.test)

(clojure.test/is (= (fib 0) 0))
(clojure.test/is (= (fib 1) 1))
(clojure.test/is (= (fib 2) 1))
(clojure.test/is (= (fib 3) 2))
(clojure.test/is (= (fib 4) 3))
(clojure.test/is (= (fib 5) 5))
(clojure.test/is (= (fib 6) 8))
(clojure.test/is (= (fib 7) 13))
(clojure.test/is (= (fib 8) 21))
(clojure.test/is (= (fib 15) 610))
(println (fib 0))
(println (fib 1))
(println (fib 2))
(println (fib 3))
(println (fib 4))
(println (fib 5))
(println (fib 6))
(println (fib 7))
(println (fib 8))
(println (fib 9))


(println "done")

File: java/src/ch1/ex119.java
package ch1;

public class ex119 {
    public static void main(String[] args) {
        Fun f = (int i) -> fib_iter(1, 0, 0, 1, i);

        System.out.println(f.apply(0) == 0);
        System.out.println(f.apply(1) == 1);
        System.out.println(f.apply(2) == 1);
        System.out.println(f.apply(3) == 2);
        System.out.println(f.apply(4) == 3);
        System.out.println(f.apply(5) == 5);
        System.out.println(f.apply(6) == 8);
        System.out.println(f.apply(15) == 610);
        System.out.println("done");
    }

    @FunctionalInterface
    interface Fun {
        int apply(int i);
    }

    static int fib_iter(int a, int b, int p, int q, int count) {
        if (count == 0) return b;
        if (even(count)) return fib_iter(a, b, p*p + q*q, q*q + 2*p*q, count / 2);
        return fib_iter(b*q + a*q + a*p, b*p + a*q, p, q, count - 1);
    }

    static boolean even(int x) {
        return x % 2 == 0;
    }
}
